restrict_download_region/restrict_region.py

```diff
@@ -0,0 +1,153 @@
+import os
+import logging
+import botocore
+import boto3
+import json
+import restrict_download_region.cfnresponse as cfnresponse
+from contextlib import contextmanager
+import requests
+from typing import List, Dict, Optional
+import itertools
+
+REGION = os.environ.get('AWS_REGION')
+POLICY_STATEMENT_ID = "DenyGetObjectForNonMatchingIp"
+SINGLE_REGION_BUCKET_TAG = 'single-region-access'
+
+
+def handler(event: dict, context: dict):
+    """
+    This lambda will either be triggered by a CloudFormation Custom Resource creation or by a recurring SNS topic
+    """
+    # context manager for the case when this lambda is triggered by aws custom resource
+    with handle_custom_resource_status_message(event, context) as custom_resource_request_type:
+        # when custom_resource_request_type is None, this lambda is being triggered by Amazon's SNS topic because IP prefixes have updated
+        buckets_to_process = [get_bucket_name_from_custom_resouce(event)] if custom_resource_request_type is not None \
+            else get_all_region_restricted_bucket_names()
+
+        region_ip_prefixes = get_ip_prefixes_for_region() if custom_resource_request_type != 'Delete' else None
+
+        s3_client = boto3.client('s3')
+
+        for bucket_name in buckets_to_process:
+            # get current bucket_policy from the s3 bucket
+            bucket_policy = get_bucket_policy(s3_client, bucket_name)
+
+            # add/update/remove ip restirction policy depending on whether region_ip_prefixes
+            process_ip_restrict_policy(bucket_name, region_ip_prefixes, custom_resource_request_type, bucket_policy)
+
+            # update with newly modified bucket policy
+            update_bucket_policy(s3_client, bucket_name, bucket_policy)
```

Comment:
this lambda looks like it's update one bucket at a time?  What happens if it fails to update the 1st bucket or one of the buckets in the middle of the list?  will it fail and not do the rest? and how would we know if it failed to process all buckets?

Reply:
True. I'll need to add a try except handling for the SNS case.